triggers/simulations/contention.rst

```diff
@@ -0,0 +1,248 @@
+.. Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+Simulation Use Case Example: Impact of Uncontrolled Contention
+==============================================================
+
+As triggers interact with the ledger asynchronously and the ledger *may* also process transactions concurrently, there is a high risk that multiple changes (e.g. exercises, archives, etc.) to the same contract might overlap each other. When this happens, contention is said to occur.
+
+In the following use-case simulations we explore some of the consequences in not controlling contention.
+
+This use-case assumes that the :doc:`development instructions <development.rst>` have been followed to setup your development and simulation environment.
+
+.. note::
+  By default, we use the open source version of Canton in our Scalatest based simulations. This version of Canton performs:
+
+  - sequential transaction processing
+  - and, due to some hard coded resource limits, is more likely to rate limit ledger command submissions.
+
+  The enterprise version of Canton does not have these limitations.
+
+Our simulation use-cases will use the Daml ``Cat`` template:
+
+.. code-block:: unused
+  template Cat
+    with
+      owner : Party
+      name : Int
+    where
+      signatory owner
+
+    choice Stroke_Cat : ()
+      controller owner
+      do
+        pure ()
+
+along with the following 3 triggers:
+
+- a breeding trigger that will create contracts
+- a culling trigger that will archive contracts
+- and a petting trigger that will exercise the ``Stroke_Cat`` choice on contracts.
+
+The culling and petting triggers are designed so that there is a high risk that they will interfere or contend with each other.
+
+The Scala simulation code is defined in :doc:`Contention.scala <scala/Contention.scala>`.
+
+Impact of Uncontrolled Contention
+---------------------------------
+
+As the ledger processes command submissions, it will potentially *lock* the contracts it is about to interact with. This allows triggers to detect `contention <https://docs.daml.com/canton/usermanual/troubleshooting_guide.html#contention>`_ as one of the following completion failure signals:
+
+- LOCAL_VERDICT_LOCKED_CONTRACTS
+- LOCAL_VERDICT_LOCKED_KEYS
+- LOCAL_VERDICT_INCONSISTENT_KEY
+- LOCAL_VERDICT_INACTIVE_CONTRACTS
+- LOCAL_VERDICT_DUPLICATE_KEY
+- CONTRACT_NOT_FOUND
+- DUPLICATE_CONTRACT_KEY
+
+If contention issues are not managed by your trigger definitions, then this can lead to:
+
+- contract exercises taking longer than expected to run (due to a need to retry the command failure)
+- contracts having longer than expected lifetimes (as archive commands might be delayed)
+- and the triggers ACS becoming larger than expected over time.
+
+When writing trigger Daml code, **it is the responsibility of the user to ensure that all potential ledger contention is managed**.
```

Comment:
Should we expend that, saying the best way to manage contention is to avoid them ? and simple pattern to avoid them ?